File: cpp/3876_uniform_array/uniform_array.cpp
#include <algorithm>
#include <cassert>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  bool uniformArray(vector<int>& nums1) {
    int mn = nums1[0];
    bool hasOdd = false;
    for (int v : nums1) {
      mn = std::min(v, mn);
      if ((v & 1) != 0) {
        hasOdd = true;
      }
    }
    if ((mn & 1) != 0) {
      return true;
    }
    return !hasOdd;
  }
};

int main() {
  vector<tuple<vector<int>, bool>> tests{
      {{1, 4, 7}, true},
      {{2, 3}, false},
      {{4, 6}, true},
  };

  for (auto& [nums1, ans] : tests) {
    assert(Solution().uniformArray(nums1) == ans);
  }
}

File: rust/3876_uniform_array/uniform_array.rs
struct Solution;

impl Solution {
    pub fn uniform_array(nums1: Vec<i32>) -> bool {
        let mut mn = nums1[0];
        let mut has_odd = false;
        for &v in &nums1 {
            if v < mn {
                mn = v;
            }
            if (v & 1) == 1 {
                has_odd = true;
            }
        }
        if (mn & 1) == 1 {
            return true;
        }
        return !has_odd;
    }
}

fn main() {
    let tests = vec![
        (vec![1, 4, 7], true),
        (vec![2, 3], false),
        (vec![4, 6], true),
    ];

    for (nums1, ans) in tests {
        assert_eq!(Solution::uniform_array(nums1), ans);
    }
}
